### PR TITLE
Fix and updated the plugin

### LINKED ISSUE #19: Unable to install latest version
I keep receiving this error with v0.2.0, both when I download the published release or build from source:

```

{"level":"error","ts":1565852540.218115,"caller":"mlog/log.go:172","msg":"Unable to activate plugin","plugin_id":"autotranslate","error":"unable to start plugin: autotranslate: Unrecognized remote plugin message: \n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol.","errorVerbose":"Unrecognized remote plugin message: \n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol.\nunable to start plugin: autotranslate\ngithub.com/mattermost/mattermost-server/plugin.(*Environment).Activate\n\t/go/src/github.com/mattermost/mattermost-server/plugin/environment.go:230\ngithub.com/mattermost/mattermost-server/app.(*App).SyncPluginsActiveState\n\t/go/src/github.com/mattermost/mattermost-server/app/plugin.go:95\ngithub.com/mattermost/mattermost-server/app.(*App).InitPlugins\n\t/go/src/github.com/mattermost/mattermost-server/app/plugin.go:126\ngithub.com/mattermost/mattermost-server/app.(*Server).RunOldAppInitalization.func6\n\t/go/src/github.com/mattermost/mattermost-server/app/server_app_adapters.go:147\ngithub.com/mattermost/mattermost-server/config.(*emitter).invokeConfigListeners.func1\n\t/go/src/github.com/mattermost/mattermost-server/config/emitter.go:35\nsync.(*Map).Range\n\t/usr/local/go/src/sync/map.go:333\ngithub.com/mattermost/mattermost-server/config.(*emitter).invokeConfigListeners\n\t/go/src/github.com/mattermost/mattermost-server/config/emitter.go:33\ngithub.com/mattermost/mattermost-server/config.(*commonStore).set\n\t/go/src/github.com/mattermost/mattermost-server/config/common.go:82\ngithub.com/mattermost/mattermost-server/config.(*FileStore).Set\n\t/go/src/github.com/mattermost/mattermost-server/config/file.go:97\ngithub.com/mattermost/mattermost-server/app.(*Server).UpdateConfig\n\t/go/src/github.com/mattermost/mattermost-server/app/config.go:53\ngithub.com/mattermost/mattermost-server/app.(*App).UpdateConfig\n\t/go/src/github.com/mattermost/mattermost-server/app/config.go:59\ngithub.com/mattermost/mattermost-server/app.(*App).EnablePlugin\n\t/go/src/github.com/mattermost/mattermost-server/app/plugin.go:245\ngithub.com/mattermost/mattermost-server/api4.enablePlugin\n\t/go/src/github.com/mattermost/mattermost-server/api4/plugin.go:191\ngithub.com/mattermost/mattermost-server/web.Handler.ServeHTTP\n\t/go/src/github.com/mattermost/mattermost-server/web/handlers.go:144\ngithub.com/mattermost/mattermost-server/vendor/github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1\n\t/go/src/github.com/mattermost/mattermost-server/vendor/github.com/NYTimes/gziphandler/gzip.go:336\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:1995\ngithub.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.(*Router).ServeHTTP\n\t/go/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux/mux.go:212\ngithub.com/mattermost/mattermost-server/app.(*RateLimiter).RateLimitHandler.func1\n\t/go/src/github.com/mattermost/mattermost-server/app/ratelimit.go:109\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:1995\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2774\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1878\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1337"}
```

### LINKED ISSUE #22: plugin does not work
When I clicked on the drop down menu, it did not work, nothing happened

image

Plugin version: 0.2.0. Revision number: 8c87f7fdf72c6436bea1774e49cc57c6cb432833
Mattermost Version: 5.16.2

### PR DESCRIPTION
**Pull request by saturninoabril**
#### Summary
- Updated minimum version to "5.22.0".
- Updated supported languages from AWS Translate.
- Translation is now supported, not only in reference to English but in any combination between source and target languages -- a _language pair_ supported by AWS Translate.
- Updated readme and code owner.
- Updated code setup - circleci setup, make file, eslint, etc.
- Updated the way it store objects. Previously it's modifying the `post` which is no longer permitted due to its immutability. Now, the store is saved as:
```
{
  ...
  plugins-autotranslate: {
    translatedPosts: {
      [post-id]: {/translation object/}
    },
    translations: {
      [translation-id]: {/translation object based on source/target language/}
    },
    userInfo: {/no change/}
  }
}
```

Test steps:
1. On initial run, `autotranslate` plugin is disabled to a user. Post a message and check that “Translate” is not available in post’s dropdown menu
2. See ephemeral help by posting `/autotranslate` or `/autotranslate help`. Check text, words or links for any discrepancy.
3. Enable the plugin for current user by `/autotranslate on` and see that it’s enabled with source: Auto and target: English.
4. Post any message in a different language [supported by AWS Translate](https://docs.aws.amazon.com/translate/latest/dg/what-is.html)
    - Check that “Translate” is available in that post. (Note: “Translate” is available to regular post only. Other post like system message should not have that option.)
    - Click “Translate” and see that the post is translated.
    - Click “close” to hide the translated message
5. Try to change different settings (info, source, target, of/off) and see if it works or translates as expected.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -4,6 +4,12 @@ go 1.12
 
 require (
 	github.com/aws/aws-sdk-go v1.26.6
-	github.com/mattermost/mattermost-server/v5 v5.18.0
-	github.com/pkg/errors v0.8.1
+	github.com/blang/semver v3.5.1+incompatible
```
Comment (hanzei) [inline go.mod]: Nit pick: `github.com/blang/semver/v4` was just released which is go modules compatible.

File: go.mod [hunk 2]
```diff
@@ -4,6 +4,12 @@ go 1.12
...
+	github.com/mattermost/ldap v3.0.4+incompatible // indirect
+	github.com/mattermost/mattermost-server/v5 v5.21.0
+	github.com/onsi/ginkgo v1.8.0 // indirect
+	github.com/onsi/gomega v1.5.0 // indirect
+	github.com/pkg/errors v0.9.1
+	github.com/stretchr/testify v1.5.1 // indirect
+	gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d // indirect
```
Comment (hanzei) [inline go.mod]: I'm wondering where these indirect dependencies come from. Maybe just remove the manually?

File: plugin.json [hunk 3]
```diff
@@ -1,8 +1,11 @@
 {
     "id": "autotranslate",
     "name": "Autotranslate",
-    "description": "Autotranslate plugin for Mattermost 5.4 and later.",
-    "version": "0.2.0",
+    "description": "Autotranslate plugin for Mattermost 5.22 and later.",
+    "homepage_url": "https://github.com/mattermost/mattermost-plugin-autotranslate",
+    "support_url": "https://github.com/mattermost/mattermost-plugin-autotranslate/issues",
```
Comment (hanzei) [inline plugin.json:6]: While touching this file let's also add:
```suggestion
    "release_notes_url": "https://github.com/mattermost/mattermost-plugin-autotranslate/releases/tag/v0.3.0",
```

File: server/activate_hooks.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package main
```
Comment (hanzei) [inline server/activate_hooks.go:1]: I'm not sure if this file is really needed. `checkServerVersion` is a relikt from the past.

File: server/configuration.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	defer p.configurationLock.Unlock()
+
+	if configuration != nil && p.configuration == configuration {
+		panic("setConfiguration called with the existing configuration")
+	}
+
+	p.configuration = configuration
+}
+
+func (p *Plugin) diffConfiguration(newConfiguration *configuration) {
```
Comment (hanzei) [inline server/configuration.go]: Is this method needed?

File: server/command.go [hunk 6]
```diff
@@ -4,83 +4,115 @@ import (
...
-	"es":   "Spanish",
-	"it":   "Italian",
-	"ja":   "Japanese",
-	"pt":   "Portuguese",
-	"ru":   "Russian",
-	"tr":   "Turkish",
+// See https://docs.aws.amazon.com/translate/latest/dg/what-is.html for updated supported languages.
+// Below is hard-coded but would be nice if AWS SDK supports getting the list programmatically
+// which is not the case currently.
+var languageCodes = map[string]interface{}{
```
Comment (hanzei) [inline server/command.go]: Why not 
```suggestion
var languageCodes = map[string]string{
```
?

**Review by hanzei — CHANGES_REQUESTED** [review]
I don't have the setup to test the plugin right now. But the changes look mostly good. I'm just curious if certain additions are really needed.

File: go.mod (hunk 1 above)
Comment (saturninoabril) [reply go.mod]: Thanks, now updated.

File: go.mod (hunk 2 above)
Comment (saturninoabril) [reply go.mod]: Thanks, now updated.

File: server/activate_hooks.go (hunk 4 above)
Comment (saturninoabril) [reply server/activate_hooks.go:1]: Okay, I'm not totally aware. Looking at other, seemed like I should be using `mattermost-plugin-api`. But if it's good, I'll leave this as is and will update per new approach on separate PR.  Any thoughts?

File: server/configuration.go (hunk 5 above)
Comment (saturninoabril) [reply server/configuration.go]: I just removed it now but let me know what's the proper approach.

**Comment by saturninoabril** [comment]
(Hmm, diff to go.sum is a bit annoying. Not happening on local. btw, I'll update all at once after dev review)

File: server/activate_hooks.go (hunk 4 above)
Comment (lieut-data) [reply server/activate_hooks.go:1]: Agree with @hanzei -- I think the only thing needed in this file is:
```
func (p *Plugin) OnActivate() error {
	if err := p.IsValid(); err != nil {
		return err
	}

	p.API.RegisterCommand(getCommand())

	return nil
}
```

**Review by lieut-data — APPROVED** [review]

File: plugin.json [hunk 7]
```diff
@@ -1,8 +1,11 @@
 {
     "id": "autotranslate",
     "name": "Autotranslate",
-    "description": "Autotranslate plugin for Mattermost 5.4 and later.",
-    "version": "0.2.0",
+    "description": "Autotranslate plugin for Mattermost 5.22 and later.",
+    "homepage_url": "https://github.com/mattermost/mattermost-plugin-autotranslate",
```
Comment (hanzei) [inline plugin.json:5]: https://github.com/mattermost/mattermost-plugin-autotranslate/pull/27#discussion_r429870396 was a bad suggestion, sorry.
```suggestion
    "homepage_url": "https://github.com/mattermost/mattermost-plugin-autotranslate",
    "support_url": "https://github.com/mattermost/mattermost-plugin-autotranslate/issues",
```

File: server/configuration.go (hunk 5 above)
Comment (hanzei) [reply server/configuration.go]: Is an unneeded method. You can just drop it

**Comment by saturninoabril** [comment]
Thanks Jesse and Ben for the reviews. PR now updated, PTAL.

**Review by ogi-m — APPROVED** [review]
Tested and working as expected 👍

**Comment by saturninoabril** [comment]
Thank you for testing Ogi!

**Review by hanzei — APPROVED** [review]
Awesome to see this plugin working again :+1:

**Pull request merged**